입출금 현황 화면 검색, 필터, 정렬 기능 구현

donggeonoh opened the pull request:
구현 사항

- Search Bar delegate textDidChange를 이용하여 검색 구현
- Segmented control을 이용하여 필터링 구현
- 각각 Button을 만들어서 정렬 기능 구현

closed #31

donggeonoh committed:
feat: 입출금 현황 화면 검색, 필터, 정렬 기능 구현

- Search Bar delegate textDidChange를 이용하여 검색 구현
- Segmented control을 이용하여 필터링 구현
- 각각 Button을 만들어서 정렬 기능 구현

donggeonoh committed:
Merge branch 'develop' into feat/31/search-deposit-withdrawal-status-and-sort-cell

donggeonoh committed:
style: 임시 테스트 관련 코드 제거 #31

HJEHA commented on BithumbYagomAcamedy/Controller/DepositWithdrawalStatusViewController.swift:
```diff
@@ -17,6 +17,11 @@ final class DepositWithdrawalStatusViewController: UIViewController {
     // MARK: - IBOutlet
     
+    @IBOutlet private weak var statusSearchBar: UISearchBar!
+    @IBOutlet weak var filterSegmentedControl: UISegmentedControl!
+    @IBOutlet weak var nameSortButton: SortButton!
+    @IBOutlet weak var depositSortButton: SortButton!
+    @IBOutlet weak var withdrawalSortButton: SortButton!
```
접근제어가 없어도 괜찮은 걸까요??:)

donggeonoh replied:
앗 이부분 빼먹었네요 ㅎㅎ 감사합니다

donggeonoh committed:
feat: SortButton 불필요한 self 삭제 #31

HJEHA commented on BithumbYagomAcamedy/Controller/DepositWithdrawalStatusViewController.swift:
```diff
@@ -71,11 +76,96 @@ final class DepositWithdrawalStatusViewController: UIViewController {
         }
     }
+    
+    private func filteredAssetsStatusesBySegmentedContorl() -> [AssetsStatus] {
+        guard let dataManager = dataManager else {
+            return []
+        }
+        let filteredData = dataManager.filterStatuses(by: .init(rawValue: filterSegmentedControl.selectedSegmentIndex))
+        
+        return filteredData
+    }
```
VC에서 이벤트처리, 뷰 그리기 이외에 연산을 하는 것 같네요:)
VC의 역할을 최소화하려면 dataManager에게 이벤트 전달만 해주고, 이벤트를 받은 dataManager는 연산(필터)을 수행한 후 VC에게
뷰를 갱신하라고 시키는게 좋을 것 같습니다:)
저 같은 경우엔 dataManager의 delegate로 VC를 두는 편 입니다:)

donggeonoh replied:
위에서 연산을 VC에서 하는 것이 아니라 데이터 매니저에서 필터링 된 데이터를 VC에서 호출하여 가져와 여러 곳에서 쓰이기 때문에(재사용성을 위해서) 메서드로 분리시켰습니다. 사실상 연산보다는 Model에게 연산 동작을 맡기고 나서 필터링 된 데이터를 가져온 후 스냅샷에 apply 하기만 하면 View에 갱신됩니다. 저 메서드가 상당히 많이 재사용 돼요 ㅎㅎ

donggeonoh replied:
해당 의견 다시 생각해보니 dataManager에서 처리해줄 수 있을 것 같아, delegate로 처리하였습니다. 훨씬 더 깔끔해졌네요 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Controller/DepositWithdrawalStatusViewController.swift:
```diff
@@ -71,11 +76,96 @@ final class DepositWithdrawalStatusViewController: UIViewController {
+            return
+        }
+        
+        guard !searchText.isEmpty else {
```
!도 좋지만 `== false`는 어떠신가요?:)

donggeonoh replied:
가독성이 더 괜찮은 것 같네요.

HJEHA commented on BithumbYagomAcamedy/Model/DepositWithdrawalStatusDataManager.swift:
```diff
@@ -9,9 +9,21 @@ import Foundation
     // MARK: - Property
     private let service: HTTPNetworkService
-    private var statuses: [AssetsStatus]
+    private(set) var statuses: [AssetsStatus]
```
VC가 이벤트 전달, 뷰 그리기 역할만 수행한다면 statuses는 private이 될 수 있을 것 같아요:)

donggeonoh replied:
저 부분 수정해보겠습니다~

HJEHA commented on BithumbYagomAcamedy/Model/DepositWithdrawalStatusDataManager.swift:
```diff
@@ -35,6 +47,60 @@ final class DepositWithdrawalStatusDataManager {
+        
+        switch type {
+        case .name:
+            sortedData = isAscend ?
+            data.sorted { $0.coinName > $1.coinName } :
+            data.sorted { $0.coinName < $1.coinName }
```
삼항 연산자를 사용하셨는데 함수로 따로 빼도 괜찮을 것 같습니다:)
`?`와 `:`가 떨어져 있어 코드를 보자마자 삼항연산자인지 판단하기 힘들 것 같아요ㅜㅜ

donggeonoh replied:
그렇다면 그냥 if else로 해야겠네요 ㅠㅠ

HJEHA commented on BithumbYagomAcamedy/View/DepositWithdrawalStatusView/SortButton.swift:
```diff
@@ -0,0 +1,63 @@
+        setTitle(title, for: .normal)
+    }
+    
+    override func awakeFromNib() {
+        commonInit()
+    }
```
`commonInit()`메서드가 init()에서도 호출되고 있는데 여러번 호출되지는 않을까요??:)

donggeonoh replied:
아 그렇군요 그러면 해당 오버라이드 메서드를 지워야겠네요 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/View/DepositWithdrawalStatusView/SortButton.swift:
```diff
@@ -0,0 +1,63 @@
+        commonInit()
+    }
+    
+    override init(frame: CGRect) {
+        self.isAscend = false
+        super.init(frame: frame)
+        commonInit()
+    }
```
인터페이스 빌더로 작업 중이라서 init(frame: )가 필요하지 않을 것 같은데 만들어주신 이유가 있을까요???:)

donggeonoh replied:
나중에 코드로 작업하는 경우도 생각했습니다~ 사실 사용을 안하니 지우는 것이 맞다고 생각이 되네요.

HJEHA commented on BithumbYagomAcamedy/View/DepositWithdrawalStatusView/SortButton.swift:
```diff
@@ -0,0 +1,63 @@
+        commonInit()
+    }
+    
+    init(title: String) {
+        self.isAscend = false
+        super.init(frame: .zero)
+        commonInit()
+        setTitle(title, for: .normal)
+    }
```
이 init(title: ) 어디서 호출되는지 알려 주실 수 있나요??:)

donggeonoh replied:
현재 사용중이지 않고있네요. 이부분도 삭제하도록 하겠습니다 ㅎㅎ

HJEHA reviewed (commented):
코멘트 확인 부탁드립니다:)

z3rosmith commented on BithumbYagomAcamedy/Model/DepositWithdrawalStatusDataManager.swift:
```diff
@@ -35,6 +47,60 @@ final class DepositWithdrawalStatusDataManager {
+        }
+    }
+    
+    func filterStatuses(by type: FilterType?) -> [AssetsStatus] {
```
statuses가 변하지 않는 만큼, `filterStatuses`에서 `filteredStatuses`가 의미전달 상  더 나을것같습니다!

donggeonoh replied:
아 좋습니다. 의견 감사합니다.

z3rosmith commented on BithumbYagomAcamedy/Model/DepositWithdrawalStatusDataManager.swift:
```diff
@@ -35,6 +47,60 @@ final class DepositWithdrawalStatusDataManager {
+        return filteredData
+    }
+    
+    func sortStatuses(data: [AssetsStatus], by type: SortType?, _ isAscend: Bool) -> [AssetsStatus] {
```
여기도 위의 댓글과 같이 statuses가 변하지 않는 만큼, `sortStatuses`에서 `sortedStatuses`가 의미전달 상  더 나을것같습니다!

donggeonoh replied:
좋습니다! 의견 감사합니다~

donggeonoh committed:
feat: DataManager delegate 추가 및 네이밍, 접근제어 변경 #31

- 입출금현황 VC button 들의 접근제어 변경
- 입출금현황에서 VC이 dataManager delegate 위임
- 입출금현황에서 dataManger에게 데이터 필터링, 정렬을 요청하면 delegate를 통해 변경하도록 수정
- dataManger의 메서드 네이밍 변경 및 접근제어 변경

donggeonoh committed:
feat: AssetsStatus 생성자 추가
